README.md
# text-to-svg

[![NPM version][npm-image]][npm-url] [![Build Status][travis-image]][travis-url]

Convert text to SVG path without native dependence.

```js
const TextToSVG = require('text-to-svg');
const textToSVG = TextToSVG.loadSync();

const attributes = {fill: 'red', stroke: 'black'};
const options = {x: 0, y: 0, fontSize: 72, anchor: 'top', attributes: attributes};

const svg = textToSVG.getSVG('hello', options);

console.log(svg);
```

```xml
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="180" height="72">
  <path fill="red" stroke="black" d="M5.27 9.28L10.62 9.28L10.62 29.36Q15.86 24.12 21.02 24.12Q26.89 24.12 29.60 29.29Q31.11 32.20 31.11 36.35L31.11 59.70L25.77 59.70L25.77 37.93Q25.77 29.21 20.18 29.21Q16.42 29.21 13.57 31.96Q10.62 34.91 10.62 38.71L10.62 59.70L5.27 59.70L5.27 9.28ZM67.68 49.08Q64.55 61.10 54.07 61.10Q47.57 61.10 43.77 55.69Q40.32 50.73 40.32 42.61Q40.32 34.84 43.56 29.88Q47.36 24.12 54 24.12Q66.97 24.12 67.82 43.70L45.74 43.70Q46.16 56.29 54.14 56.29Q60.47 56.29 62.05 49.08L67.68 49.08M62.05 39.09Q60.89 28.93 54 28.93Q47.36 28.93 45.95 39.09L62.05 39.09ZM92.81 51.82Q92.81 54.91 95.77 54.91Q98.19 54.91 101.07 54.35L101.07 59.73Q96.82 60.33 94.82 60.33Q87.19 60.33 87.19 52.84L87.19 9.28L92.81 9.28L92.81 51.82ZM128.81 51.82Q128.81 54.91 131.77 54.91Q134.19 54.91 137.07 54.35L137.07 59.73Q132.82 60.33 130.82 60.33Q123.19 60.33 123.19 52.84L123.19 9.28L128.81 9.28L128.81 51.82ZM162.07 24.12Q168.68 24.12 172.44 29.95Q175.68 34.80 175.68 42.61Q175.68 48.48 173.74 52.91Q170.16 61.14 161.93 61.14Q155.57 61.14 151.77 55.72Q148.32 50.77 148.32 42.61Q148.32 33.82 152.30 28.79Q156.09 24.12 162.07 24.12M161.93 29.14Q158.06 29.14 155.88 33.19Q153.95 36.74 153.95 42.61Q153.95 48.02 155.53 51.43Q157.71 56.11 162 56.11Q165.94 56.11 168.12 52.07Q170.05 48.52 170.05 42.68Q170.05 36.60 168.05 33.12Q165.90 29.14 161.93 29.14Z"/>
</svg>
```

## Installation

```
$ npm install --save text-to-svg
```

## Constructor

An example for loading default font synchronously. The default font is [IPA font](http://ipafont.ipa.go.jp/). ** This method only works on Node.js. **

```js
const textToSVG = TextToSVG.loadSync();
const svg = textToSVG.getSVG('hello');
console.log(svg);
```

An example for loading font synchronously. ** This method only works on Node.js. **

```js
// Argument is file path (NOT URL)
const textToSVG = TextToSVG.loadSync('/fonts/Noto-Sans.otf');
const svg = textToSVG.getSVG('hello');
console.log(svg);
```

An example for loading font asynchronously.

```js
// First argument is URL on web browsers, but it is file path on Node.js.
TextToSVG.load('/fonts/Noto-Sans.otf', function(err, textToSVG) {
    const svg = textToSVG.getSVG('hello');
    console.log(svg);
});
```

## API

### `TextToSVG.getD(text, options = {})`

Get the path data for `d` attribute of `path`.

 - `text`: Text to convert to SVG path.

Options is an optional object containing:

 - `x`: Horizontal position of the beginning of the text. (default: `0`)
 - `y`: Vertical position of the baseline of the text. (default: `0`)
 - `fontSize`: Size of the text (default: `72`).
 - `kerning`: if `true` takes kerning information into account (default: `true`)
 - `letterSpacing`: letter-spacing value in `em`.
 - `tracking`: tracking value. (em / 1000)
 - `anchor`: Anchor of object in coordinate. (default: `'left baseline'`)
   - (`left`, `center`, `right`) + (`baseline`, `top`, `middle`, `bottom`)

Return value example

```
M5.27-54.07L10.62-54.07L10.62-34.00Q15.86-39.23 21.02-39.23Q26.89-39.23 29.60-34.07Q31.11-31.15 31.11-27L31.11-3.66L25.77-3.66L25.77-25.42Q25.77-34.14 20.18-34.14Q16.42-34.14 13.57-31.39Q10.62-28.44 10.62-24.64L10.62-3.66L5.27-3.66L5.27-54.07ZM67.68-14.27Q64.55-2.25 54.07-2.25Q47.57-2.25 43.77-7.66Q40.32-12.62 40.32-20.74Q40.32-28.51 43.56-33.47Q47.36-39.23 54-39.23Q66.97-39.23 67.82-19.65L45.74-19.65Q46.16-7.07 54.14-7.07Q60.47-7.07 62.05-14.27L67.68-14.27M62.05-24.26Q60.89-34.42 54-34.42Q47.36-34.42 45.95-24.26L62.05-24.26ZM92.81-11.53Q92.81-8.44 95.77-8.44Q98.19-8.44 101.07-9L101.07-3.62Q96.82-3.02 94.82-3.02Q87.19-3.02 87.19-10.51L87.19-54.07L92.81-54.07L92.81-11.53ZM128.81-11.53Q128.81-8.44 131.77-8.44Q134.19-8.44 137.07-9L137.07-3.62Q132.82-3.02 130.82-3.02Q123.19-3.02 123.19-10.51L123.19-54.07L128.81-54.07L128.81-11.53ZM162.07-39.23Q168.68-39.23 172.44-33.40Q175.68-28.55 175.68-20.74Q175.68-14.87 173.74-10.44Q170.16-2.21 161.93-2.21Q155.57-2.21 151.77-7.63Q148.32-12.59 148.32-20.74Q148.32-29.53 152.30-34.56Q156.09-39.23 162.07-39.23M161.93-34.21Q158.06-34.21 155.88-30.16Q153.95-26.61 153.95-20.74Q153.95-15.33 155.53-11.92Q157.71-7.24 162-7.24Q165.94-7.24 168.12-11.29Q170.05-14.84 170.05-20.67Q170.05-26.75 168.05-30.23Q165.90-34.21 161.93-34.21Z
```

### `TextToSVG.getPath(text, options = {})`

Get the `path` element of SVG.

 - `text`: Text to convert to SVG path.

Options is an optional object containing:

 - `x`: Horizontal position of the beginning of the text. (default: `0`)
 - `y`: Vertical position of the baseline of the text. (default: `0`)
 - `fontSize`: Size of the text (default: `72`).
 - `kerning`: if `true` takes kerning information into account (default: `true`)
 - `letterSpacing`: letter-spacing value in `em`.
 - `tracking`: tracking value. (em / 1000)
 - `anchor`: Anchor of object in coordinate. (default: `'left baseline'`)
   - (`left`, `center`, `right`) + (`baseline`, `top`, `middle`, `bottom`)
 - `attributes`: Key-Value pairs of attributes of `path` element.

Return value example

```xml
<path fill="red" stroke="black" d="M5.27-54.07L10.62-54.07L10.62-34.00Q15.86-39.23 21.02-39.23Q26.89-39.23 29.60-34.07Q31.11-31.15 31.11-27L31.11-3.66L25.77-3.66L25.77-25.42Q25.77-34.14 20.18-34.14Q16.42-34.14 13.57-31.39Q10.62-28.44 10.62-24.64L10.62-3.66L5.27-3.66L5.27-54.07ZM67.68-14.27Q64.55-2.25 54.07-2.25Q47.57-2.25 43.77-7.66Q40.32-12.62 40.32-20.74Q40.32-28.51 43.56-33.47Q47.36-39.23 54-39.23Q66.97-39.23 67.82-19.65L45.74-19.65Q46.16-7.07 54.14-7.07Q60.47-7.07 62.05-14.27L67.68-14.27M62.05-24.26Q60.89-34.42 54-34.42Q47.36-34.42 45.95-24.26L62.05-24.26ZM92.81-11.53Q92.81-8.44 95.77-8.44Q98.19-8.44 101.07-9L101.07-3.62Q96.82-3.02 94.82-3.02Q87.19-3.02 87.19-10.51L87.19-54.07L92.81-54.07L92.81-11.53ZM128.81-11.53Q128.81-8.44 131.77-8.44Q134.19-8.44 137.07-9L137.07-3.62Q132.82-3.02 130.82-3.02Q123.19-3.02 123.19-10.51L123.19-54.07L128.81-54.07L128.81-11.53ZM162.07-39.23Q168.68-39.23 172.44-33.40Q175.68-28.55 175.68-20.74Q175.68-14.87 173.74-10.44Q170.16-2.21 161.93-2.21Q155.57-2.21 151.77-7.63Q148.32-12.59 148.32-20.74Q148.32-29.53 152.30-34.56Q156.09-39.23 162.07-39.23M161.93-34.21Q158.06-34.21 155.88-30.16Q153.95-26.61 153.95-20.74Q153.95-15.33 155.53-11.92Q157.71-7.24 162-7.24Q165.94-7.24 168.12-11.29Q170.05-14.84 170.05-20.67Q170.05-26.75 168.05-30.23Q165.90-34.21 161.93-34.21Z"/>
```

### `TextToSVG.getSVG(text, options = {})`

Get the SVG.

 - `text`: Text to convert to SVG path.

Options is an optional object containing:

 - `x`: Horizontal position of the beginning of the text. (default: `0`)
 - `y`: Vertical position of the baseline of the text. (default: `0`)
 - `fontSize`: Size of the text (default: `72`).
 - `kerning`: if `true` takes kerning information into account (default: `true`)
 - `letterSpacing`: letter-spacing value in `em`.
 - `tracking`: tracking value. (em / 1000)
 - `anchor`: Anchor of object in coordinate. (default: `'left baseline'`)
   - (`left`, `center`, `right`) + (`baseline`, `top`, `middle`, `bottom`)
 - `attributes`: Key-Value pairs of attributes of `path` element.

Return value example

```xml
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path fill="red" stroke="black" d="M5.27-54.07L10.62-54.07L10.62-34.00Q15.86-39.23 21.02-39.23Q26.89-39.23 29.60-34.07Q31.11-31.15 31.11-27L31.11-3.66L25.77-3.66L25.77-25.42Q25.77-34.14 20.18-34.14Q16.42-34.14 13.57-31.39Q10.62-28.44 10.62-24.64L10.62-3.66L5.27-3.66L5.27-54.07ZM67.68-14.27Q64.55-2.25 54.07-2.25Q47.57-2.25 43.77-7.66Q40.32-12.62 40.32-20.74Q40.32-28.51 43.56-33.47Q47.36-39.23 54-39.23Q66.97-39.23 67.82-19.65L45.74-19.65Q46.16-7.07 54.14-7.07Q60.47-7.07 62.05-14.27L67.68-14.27M62.05-24.26Q60.89-34.42 54-34.42Q47.36-34.42 45.95-24.26L62.05-24.26ZM92.81-11.53Q92.81-8.44 95.77-8.44Q98.19-8.44 101.07-9L101.07-3.62Q96.82-3.02 94.82-3.02Q87.19-3.02 87.19-10.51L87.19-54.07L92.81-54.07L92.81-11.53ZM128.81-11.53Q128.81-8.44 131.77-8.44Q134.19-8.44 137.07-9L137.07-3.62Q132.82-3.02 130.82-3.02Q123.19-3.02 123.19-10.51L123.19-54.07L128.81-54.07L128.81-11.53ZM162.07-39.23Q168.68-39.23 172.44-33.40Q175.68-28.55 175.68-20.74Q175.68-14.87 173.74-10.44Q170.16-2.21 161.93-2.21Q155.57-2.21 151.77-7.63Q148.32-12.59 148.32-20.74Q148.32-29.53 152.30-34.56Q156.09-39.23 162.07-39.23M161.93-34.21Q158.06-34.21 155.88-30.16Q153.95-26.61 153.95-20.74Q153.95-15.33 155.53-11.92Q157.71-7.24 162-7.24Q165.94-7.24 168.12-11.29Q170.05-14.84 170.05-20.67Q170.05-26.75 168.05-30.23Q165.90-34.21 161.93-34.21Z"/></svg>
```

### `TextToSVG.getMetrics(text, option = {})`

Measure text size.

 - `text`: Text to measure size.

Options is an optional object containing:

 - `x`: Horizontal position of the beginning of the text. (default: `0`)
 - `y`: Vertical position of the baseline of the text. (default: `0`)
 - `fontSize`: Size of the text (default: `72`).
 - `kerning`: if `true` takes kerning information into account (default: `true`)
 - `letterSpacing`: letter-spacing value in `em`.
 - `tracking`: tracking value. (em / 1000)
 - `anchor`: Anchor of object in coordinate. (default: `'left baseline'`)

An example of return value.

```json
{
  "x": 0,
  "y": -63.3515625,
  "baseline": 0,
  "width": 180,
  "height": 72,
  "ascender": 63.3515625,
  "descender": -8.6484375
}
```

## License

MIT

## Credits

text-to-svg depends on the following softwares. I thank great authors a lot.

- [opentype.js](https://github.com/nodebox/opentype.js): Copyright (c) 2015 Frederik De Bleser
- [commander](https://github.com/tj/commander.js): Copyright (c) 2011 TJ Holowaychuk <tj@vision-media.ca>

These are released under the [MIT license](https://opensource.org/licenses/MIT)

[npm-url]: https://npmjs.org/package/text-to-svg
[npm-image]: https://badge.fury.io/js/text-to-svg.svg
[travis-url]: https://travis-ci.org/shrhdk/text-to-svg
[travis-image]: https://travis-ci.org/shrhdk/text-to-svg.svg?branch=master
[gitter-url]: https://gitter.im/shrhdk/text-to-svg
[gitter-image]: https://badges.gitter.im/Join%20Chat.svg


CHANGELOG.md
# CHANGELOG

## 3.1.5

 - Fix deployment script.

## ~~3.1.4~~

**This version had been failed to deploy. So do not use this version.**

 - Tested on Node.js v10.15.2
 - Update dependencies.
   - `opentype.js`: `0.7.3` -> `0.11.0`

## 3.1.3

 - Tested on Node.js v6.11.1
 - Update dependencies.
   - `opentype.js`: `0.7.1` -> `0.7.3`

## 3.1.2

 - Pin opentype.js version to `0.7.1`. (see #25)

## 3.1.1

 - Tested on Node.js v6.10.3
 - Update dependencies.
   - `opentype.js`: `0.6.6` -> `0.7.1`

## 3.1.0

 - Support letter-spacing option.
 - Support tracking option.

## 3.0.1

 - Fix #10

## 3.0.0

 - Simplify module layout
   - Please use `const TextToSVG = require('text-to-svg');`<br>instead of `const TextToSVG = require('text-to-svg').TextToSVG;`
 - Add `loadSync` method.
   - Please use `const textToSVG = TextToSVG.loadSync();`<br>instead of `const textToSVG = new TextToSVG();`
 - Add `load` method for asynchronous font loading.
 - Add `getMetrics` method instead of `getSize` method.
 - Correct vertical anchor option.
   - Correct `top` and `bottom`.
   - Add `baseline`.
 - Add `width` and `height` attribute to SVG element.

## 2.2.2

 - Tested on Node.js v5.3.0
 - Update dependencies.
   - `opentype.js`: `0.5.0` -> `0.6.0`

## 2.2.1

 - Tested on Node.js v4.2.1
 - Update dependencies.
 - Command line tool shows help when args are not provided.
 - Fix API document.

## 2.2.0

 - Add `getSize` method.
 - Add `anchor` option.
   - `left`, `center` , `right`, `top`, `middle`, `bottom`
 - Add command line tool.

## 2.1.0

 - Add `anchor` option.

## 2.0.2

 - Fix index.js

## 2.0.1

 - Include index.js to npm files.

## 2.0.0

 - Reform the API.
 - Include default font.
 - Remove the command line tool.

## 1.1.0

 - Add TextToSVG.getD(text, x, y, fontSize, options);

## 1.0.0

initial release


test/browser.js
/**
 * Copyright (c) 2016 Hideki Shiro
 */

/* eslint-disable no-undef */

import TextToSVG from '../src';

window.addEventListener('load', () => {
  const fontURLInput = document.querySelector('#font-url-input');
  const loadButton = document.querySelector('#load-button');
  const loadStatus = document.querySelector('#load-status');
  const textInput = document.querySelector('#text-input');
  const xInput = document.querySelector('#x-input');
  const yInput = document.querySelector('#y-input');
  const fontSizeInput = document.querySelector('#font-size-input');
  const kerningInput = document.querySelector('#kerning-input');
  const anchorInput = document.querySelector('#anchor-input');
  const getSVGButton = document.querySelector('#get-svg-button');
  const debugInput = document.querySelector('#debug-input');
  const result = document.querySelector('#result');

  let textToSVG;

  loadButton.onclick = () => {
    loadStatus.textContent = 'loading...';
    const url = fontURLInput.value;
    TextToSVG.load(url, (err, t2s) => {
      if (err) {
        loadStatus.textContent = err;
        return;
      }

      loadStatus.textContent = 'success!';
      textToSVG = t2s;
    });
  };

  getSVGButton.onclick = () => {
    if (textToSVG === null) {
      alert('load font first');
      return;
    }

    const text = textInput.value || '';
    const x = Number(xInput.value || 0);
    const y = Number(yInput.value || 0);
    const fontSize = Number(fontSizeInput.value || 72);
    const kerning = kerningInput.checked;
    const anchor = anchorInput.value || '';

    let svg;
    if (debugInput.checked) {
      svg = textToSVG.getDebugSVG(text, { x, y, fontSize, kerning, anchor });
    } else {
      svg = textToSVG.getSVG(text, { x, y, fontSize, kerning, anchor });
    }

    result.innerHTML = svg;
  };
});


test/html-reporter.js
/**
 * Copyright (c) 2016 Hideki Shiro
 */

/* eslint-disable no-console,  */

import fs from 'fs';
import TextToSVG from '../src';
const textToSVG = TextToSVG.loadSync();

const STYLE = `
  table {
    border-collapse: collapse;
  }

  tr.pass {
    background-color: #90EE90;
  }

  tr.fail {
    background-color: #FFC0CB;
  }

  th, td {
    padding: 3px;
    border: 1px solid #b9b9b9;
  }
`;

export default function(runner, mochaOptions) {
  let failures = 0;
  const stack = []; // 0:'TextToSVG' <- 1:'method' <- 2:'text' = [SP]

  const dest = mochaOptions.reporterOptions.dest;
  console.log(`write test report to ${dest}`);
  fs.writeFileSync(dest, '');

  function write(str) {
    fs.appendFileSync(dest, str);
  }

  function ontest(test, err) {
    const text = stack[2];
    const options = JSON.parse(test.title);
    const metrics = textToSVG.getMetrics(text, options);
    const d = textToSVG.getD(text, options);
    const svg = textToSVG.getDebugSVG(text, options);

    if (err) {
      failures++;
    }

    write(`<tr class="${err ? 'fail' : 'pass'}">`);
    write(`<td><pre>${JSON.stringify(options, null, 2)}</pre></td>`);
    write(`<td>${svg}</td>`);
    write(`<td><pre>${JSON.stringify(metrics, null, 2)}</pre></td>`);
    write(`<td><pre>${d}</pre></td>`);
    write('</tr>');
  }

  runner.on('start', () => {
    write('<html >');
    write('<head>');
    write('<meta charset="UTF-8" />');
    write('<title>TextToSVG</title>');
    write('</head>');
    write('<body>');
    write(`<style>${STYLE}</style>`);
  });

  runner.on('suite', suite => {
    if (suite.root) {
      return;
    }

    stack.push(suite.title);
    switch (stack.length - 1) {
      case 0: // TextToSVG
        write(`<h1>${suite.title}</h1>`);
        break;
      case 1: // method
        write(`<h2>${suite.title}</h2>`);
        break;
      case 2: // text
        write(`<h3>${suite.title}</h3>`);
        write('<table>');
        write('<tr>');
        write('<th>options</th>');
        write('<th>svg</th>');
        write('<th>getMetrics</th>');
        write('<th>getD</th>');
        write('</tr>');
        break;
      default:
        throw new Error('Unknown Depth');
    }
  });

  runner.on('suite end', suite => {
    if (stack.length === 3) {
      write('</table>');
    }
    stack.pop(suite.title);
  });

  runner.on('pass', (test) => {
    ontest(test, null);
  });

  runner.on('fail', (test, err) => {
    ontest(test, err);
  });

  runner.on('end', () => {
    write('</body></html>');
    process.exit(failures);
  });
}

module.exports = exports.default;


gulpfile.js
var assert = require('assert');
var gulp = require('gulp');
var del = require('del');
var babel = require('gulp-babel');
var mocha = require('gulp-mocha');

// Clean

gulp.task('clean', function (done) {
  return del(['build', 'public/lib', 'npm-debug.log', '!*/.gitkeep'], done);
});

// Build

gulp.task('build:src', function () {
  return gulp.src('src/**/*.js')
    .pipe(babel({presets: ['es2015']}))
    .pipe(gulp.dest('build/src/'));
});

gulp.task('build:res', function () {
  return gulp.src('fonts/**/*')
    .pipe(gulp.dest('build/fonts/'));
});

gulp.task('build:test:src', function () {
  return gulp.src('test/**/*.js')
    .pipe(babel())
    .pipe(gulp.dest('build/test/'));
});

gulp.task('build', ['build:src', 'build:res', 'build:test:src']);

// Test

gulp.task('version-check', function() {
  var packageVer = require('./package.json')['version'];
  var tagVer = process.env['TRAVIS_TAG'];

  if(tagVer !== '') {
    assert.equal(packageVer, tagVer, 'Package version and tagged version are mismatched.');
  }
});

gulp.task('test', ['build', 'version-check'], function (done) {
  gulp.src('build/test/**/*.js')
    .pipe(mocha())
    .on('end', done);
});


gulpfile.babel.js
/**
 * Copyright (c) 2016 Hideki Shiro
 */

/* eslint-disable no-process-env, global-require */

import path from 'path';
import assert from 'assert';
import gulp from 'gulp';
import del from 'del';
import babel from 'gulp-babel';
import mocha from 'gulp-mocha';
import eslint from 'gulp-eslint';
import browserify from 'browserify';
import source from 'vinyl-source-stream';

// Clean

gulp.task('clean', done => {
  return del(['build', 'npm-debug.log', '!*/.gitkeep'], done);
});

// Lint

gulp.task('lint', () => {
  return gulp.src(['src/**/*.js', 'test/**/*.js', 'gulpfile.babel.js'])
    .pipe(eslint())
    .pipe(eslint.format())
    .pipe(eslint.failOnError());
});

// Build

gulp.task('build:src', () => {
  return gulp.src('src/**/*.js')
    .pipe(babel())
    .pipe(gulp.dest('build/src/'));
});

gulp.task('build:res', () => {
  return gulp.src('fonts/**/*')
    .pipe(gulp.dest('build/fonts/'));
});

gulp.task('build:test:src', () => {
  return gulp.src('test/**/*.js')
    .pipe(babel())
    .pipe(gulp.dest('build/test/'));
});

gulp.task('build:test:browser', () => {
  return browserify('./build/test/browser.js')
    .bundle()
    .pipe(source('browser.js'))
    .pipe(gulp.dest('build/test'));
});

gulp.task('build', gulp.series('build:src', 'build:res', 'build:test:src', 'build:test:browser'));

// Test

gulp.task('version-check', done => {
  const packageVer = require('./package.json').version;
  const tagVer = process.env.TRAVIS_TAG;
  if (tagVer) {
    assert.equal(packageVer, tagVer, `Package version and tagged version are mismatched. Package version is ${packageVer}, but tagged version is ${tagVer}`);
  }
  done();
});

gulp.task('test', gulp.series(gulp.parallel('build', 'lint', 'version-check'), () => {
  return gulp.src('build/test/index.js')
    .pipe(mocha());
}));

gulp.task('test:html', gulp.series(gulp.parallel('build', 'lint', 'version-check'), () => {
  const reporter = path.join(__dirname, './build/test/html-reporter.js');
  const dest = path.join(__dirname, './build/test/result.html');
  return gulp.src('build/test/index.js')
    .pipe(mocha({ reporter, reporterOptions: { dest } }));
}));


src/index.js
/**
 * Copyright (c) 2016 Hideki Shiro
 */

import * as opentype from 'opentype.js';

const DEFAULT_FONT = require('path').join(__dirname, '../fonts/ipag.ttf');

// Private method

function parseAnchorOption(anchor) {
  let horizontal = anchor.match(/left|center|right/gi) || [];
  horizontal = horizontal.length === 0 ? 'left' : horizontal[0];

  let vertical = anchor.match(/baseline|top|bottom|middle/gi) || [];
  vertical = vertical.length === 0 ? 'baseline' : vertical[0];

  return { horizontal, vertical };
}

export default class TextToSVG {
  constructor(font) {
    this.font = font;
  }

  static loadSync(file = DEFAULT_FONT) {
    return new TextToSVG(opentype.loadSync(file));
  }

  static load(url, cb) {
    opentype.load(url, (err, font) => {
      if (err !== null) {
        return cb(err, null);
      }

      return cb(null, new TextToSVG(font));
    });
  }

  getWidth(text, options) {
    const fontSize = options.fontSize || 72;
    const kerning = 'kerning' in options ? options.kerning : true;
    const fontScale = 1 / this.font.unitsPerEm * fontSize;

    let width = 0;
    const glyphs = this.font.stringToGlyphs(text);
    for (let i = 0; i < glyphs.length; i++) {
      const glyph = glyphs[i];

      if (glyph.advanceWidth) {
        width += glyph.advanceWidth * fontScale;
      }

      if (kerning && i < glyphs.length - 1) {
        const kerningValue = this.font.getKerningValue(glyph, glyphs[i + 1]);
        width += kerningValue * fontScale;
      }

      if (options.letterSpacing) {
        width += options.letterSpacing * fontSize;
      } else if (options.tracking) {
        width += (options.tracking / 1000) * fontSize;
      }
    }
    return width;
  }

  getHeight(fontSize) {
    const fontScale = 1 / this.font.unitsPerEm * fontSize;
    return (this.font.ascender - this.font.descender) * fontScale;
  }

  getMetrics(text, options = {}) {
    const fontSize = options.fontSize || 72;
    const anchor = parseAnchorOption(options.anchor || '');

    let width = this.getWidth(text, options);
    if (options.shadow) {
      width += 20;
    }
    const height = this.getHeight(fontSize);

    const fontScale = 1 / this.font.unitsPerEm * fontSize;
    const ascender = this.font.ascender * fontScale;
    const descender = this.font.descender * fontScale;

    let x = options.x || 0;
    switch (anchor.horizontal) {
      case 'left':
        x -= 0;
        break;
      case 'center':
        x -= width / 2;
        break;
      case 'right':
        x -= width;
        break;
      default:
        throw new Error(`Unknown anchor option: ${anchor.horizontal}`);
    }

    let y = options.y || 0;
    switch (anchor.vertical) {
      case 'baseline':
        y -= ascender;
        break;
      case 'top':
        y -= 0;
        break;
      case 'middle':
        y -= height / 2;
        break;
      case 'bottom':
        y -= height;
        break;
      default:
        throw new Error(`Unknown anchor option: ${anchor.vertical}`);
    }

    const baseline = y + ascender;

    return {
      x,
      y,
      baseline,
      width,
      height,
      ascender,
      descender,
    };
  }

  getD(text, options = {}) {
    const fontSize = options.fontSize || 72;
    const kerning = 'kerning' in options ? options.kerning : true;
    const letterSpacing = 'letterSpacing' in options ? options.letterSpacing : false;
    const tracking = 'tracking' in options ? options.tracking : false;
    const metrics = this.getMetrics(text, options);
    const path = this.font.getPath(text, metrics.x, metrics.baseline, fontSize, { kerning, letterSpacing, tracking });

    return path.toPathData();
  }

  getPath(text, options = {}) {
    const attributes = Object.keys(options.attributes || {})
      .map(key => `${key}="${options.attributes[key]}"`)
      .join(' ');
    const d = this.getD(text, options);

    if (attributes) {
      return `<path ${attributes} filter="url(#f3)" d="${d}"/>`;
    }

    return `<path filter="url(#f3)" d="${d}"/>`;
  }

  getSVG(text, options = {}) {
    const metrics = this.getMetrics(text, options);
    let svg = `<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="${metrics.width}" height="${metrics.height}">`;

    // hard coded the offsets for me
    // make sure to set x="-10%" if you want to see shadow on the left of the original text
    if (options.shadow) {
      svg += `<defs><filter id="f3" x="-10%" y="0" width="200%" height="200%">`;
      svg += `<feOffset result="offOut" in="SourceAlpha" dx="2" dy="3" />`;
      svg += `<feGaussianBlur result="blurOut" in="offOut" stdDeviation="4" />`;
      svg += `<feBlend in="SourceGraphic" in2="blurOut" mode="normal" />`;
      svg += `</filter></defs>`;      
    }

    svg += this.getPath(text, options);
    svg += '</svg>';

    return svg;
  }

  getDebugSVG(text, options = {}) {
    options = JSON.parse(JSON.stringify(options));

    options.x = options.x || 0;
    options.y = options.y || 0;
    const metrics = this.getMetrics(text, options);
    const box = {
      width: Math.max(metrics.x + metrics.width, 0) - Math.min(metrics.x, 0),
      height: Math.max(metrics.y + metrics.height, 0) - Math.min(metrics.y, 0),
    };
    const origin = {
      x: box.width - Math.max(metrics.x + metrics.width, 0),
      y: box.height - Math.max(metrics.y + metrics.height, 0),
    };

    // Shift text based on origin
    options.x += origin.x;
    options.y += origin.y;

    let svg = `<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="${box.width}" height="${box.height}">`;
    svg += `<path fill="none" stroke="red" stroke-width="1" d="M0,${origin.y}L${box.width},${origin.y}"/>`; // X Axis
    svg += `<path fill="none" stroke="red" stroke-width="1" d="M${origin.x},0L${origin.x},${box.height}"/>`; // Y Axis
    svg += this.getPath(text, options);
    svg += '</svg>';

    return svg;
  }
}

module.exports = exports.default;
